Makefile

```diff
@@ -1,55 +1,15 @@
 # simple makefile to simplify repetitive build env management tasks under posix
 
-# caution: testing won't work on windows, see README
-
 PYTHON ?= python
-CYTHON ?= cython
-PYTEST ?= pytest
-
-# skip doctests on 32bit python
-BITS := $(shell python -c 'import struct; print(8 * struct.calcsize("P"))')
-
-all: clean inplace test
 
-clean:
+clean-setuptools:
 	$(PYTHON) setup.py clean
 	rm -rf dist
 
-in: inplace # just a shortcut
-inplace:
+inplace-setuptools:
 	$(PYTHON) setup.py build_ext -i
 
+dev: dev-meson
+
 dev-meson:
 	pip install --verbose --no-build-isolation --editable . --check-build-dependencies --config-settings editable-verbose=true
-
-clean-meson:
-	pip uninstall -y scikit-learn
-
```

Comment:
Why not keeping this one ? Also, is it enough ? You don't need a `git clean` ?

Reply:
Also, since I need to run the commands manually on windows, I find it nice to just have to look into the Makefile to find them

Reply:
I can keep it if you want, basically you need this if you switch between setuptools and Meson.

If you only use Meson, you do once `make dev-meson` and then importing `sklearn` will recompile as needed at import-time. That means, you never have to think do I need to build or not.

Also I find that since scikit-learn is installed as a package that if I want to remove it, `pip uninstall scikit-learn -y` is something that most people know how to do.

Reply:
With Meson (or rather ninja) in principle you never need to clean. In the case I wanted to make 100% sure there was not something weird happening, I did `rm -rf build` because that's where `pip install -e .` will put stuff by default something like `build/cp312` for a Python 3.12 install.

Reply:
Actually this is broken:

```
$ make dev
...
$ git clean -xdf
$ python -c "import sklearn; sklearn.show_versions()"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1138, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1078, in _find_spec
  File "/Users/ogrisel/miniforge3/envs/dev/lib/python3.11/site-packages/_scikit_learn_editable_loader.py", line 310, in find_spec
    tree = self._rebuild()
           ^^^^^^^^^^^^^^^
  File "/Users/ogrisel/miniforge3/envs/dev/lib/python3.11/site-packages/_scikit_learn_editable_loader.py", line 338, in _rebuild
    if self._work_to_do(env):
       ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ogrisel/miniforge3/envs/dev/lib/python3.11/site-packages/_scikit_learn_editable_loader.py", line 325, in _work_to_do
    p = subprocess.run(dry_run_build_cmd, cwd=self._build_path, env=env, capture_output=True)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ogrisel/miniforge3/envs/dev/lib/python3.11/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ogrisel/miniforge3/envs/dev/lib/python3.11/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/ogrisel/miniforge3/envs/dev/lib/python3.11/subprocess.py", line 1950, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/Users/ogrisel/code/scikit-learn/build/cp311'
```

Ideally one would want meson-python to tolerate that and rebuild the missing build directory, no?

Reply:
I tried:

```
$ mkdir build/cp311
```

but then:

```
$ python -c "import sklearn; sklearn.show_versions()"
meson-python: building scikit-learn: /Users/ogrisel/miniforge3/envs/dev/bin/ninja
ninja: error: loading 'build.ninja': No such file or directory
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1138, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1078, in _find_spec
  File "/Users/ogrisel/miniforge3/envs/dev/lib/python3.11/site-packages/_scikit_learn_editable_loader.py", line 310, in find_spec
    tree = self._rebuild()
           ^^^^^^^^^^^^^^^
  File "/Users/ogrisel/miniforge3/envs/dev/lib/python3.11/site-packages/_scikit_learn_editable_loader.py", line 346, in _rebuild
    with open(install_plan_path, 'r', encoding='utf8') as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/Users/ogrisel/code/scikit-learn/build/cp311/meson-info/intro-install_plan.json'
```

Reply:
So as long as this is broken, I would do:

```
clean: clean-meson

clean-meson:
    rm -rf build
    pip uninstall -y scikit-learn

Reply:
+1

Reply:
:thinking: I probably was not clear in my previous comment before, here is what I wanted to say:
- if you are only working with Meson, you do `make dev-meson` once and then use scikit-
learn since importing will retrigger compilation. I guess people who have switched to Meson probably never had to think much about "cleaning", it just works
- `pip uninstall -y scikit-learn` is the way to "clean" Meson, in case you want to switch to setuptools. I am guessing most people don't have to do this unless you switch to an old PR.
- every once in a while, like every month, I am looking at a fishy behaviour, and just because I am paranoid I do `rm -rf build`. This is clearly not needed and also not a good idea to do in general because you may have other build there of scikit-learn built for other other environments and Python versions.

All in all if you insist you want a `make clean` and `make clean-meson` I would stick with the previous behaviour of only doing `pip uninstall -y scikit-learn`

Reply:
I think it's helpful to have an official way to handle the cleaning of a dev build that does not result in one of the broken states I posted above.

Because indeed some people are paranoid and will attempt to do a "clean" operation manually from time to time (even if it's useless, just in case) and end up in the same state as I did.

Reply:
> Because indeed some people are paranoid and will attempt to do a "clean" operation manually from time to time

I'm one of those guys and I'm pretty sure it got me out of a very broken state sometimes :)
Anyway, even if it trivial, I'd rather have the short cut, and I like being able to see all the necessary commands explicitly at one place.

Reply:
OK is my last commit good enough or you want an additional `clean-paranoid` target that does `rm -rf build`

Reply:
good enough for me